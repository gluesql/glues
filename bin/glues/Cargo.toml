[package]
name = "glues"
version.workspace = true
edition.workspace = true
license.workspace = true
repository.workspace = true
authors.workspace = true
description = "Unified CLI entry point for the Glues toolkit"

[dependencies]
clap = { version = "4.5.4", features = ["derive"] }
color-eyre = "0.6.3"
glues-tui = { path = "../../tui", package = "glues-tui", default-features = false, features = ["cli"] }
glues-server = { path = "../../server" }
tokio = { version = "1.41.0", features = ["macros", "rt-multi-thread"] }

[features]
default = []
