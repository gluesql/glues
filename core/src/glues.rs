use {
    crate::{
        Event, Result, Transition,
        backend::CoreBackend,
        state::{EntryState, State},
        task::{Task, handle_tasks},
    },
    std::{
        collections::VecDeque,
        sync::{
            Arc, Mutex,
            mpsc::{Sender, channel},
        },
        thread::JoinHandle,
    },
};

pub struct Glues {
    pub db: Option<Box<dyn CoreBackend + Send>>,
    pub state: State,

    pub task_tx: Sender<Task>,
    pub task_handle: JoinHandle<()>,
    pub transition_queue: Arc<Mutex<VecDeque<Transition>>>,
}

impl Default for Glues {
    fn default() -> Self {
        Self::new()
    }
}

impl Glues {
    pub fn new() -> Self {
        let transition_queue = Arc::new(Mutex::new(VecDeque::new()));
        let (task_tx, task_rx) = channel();
        let task_handle = handle_tasks(task_rx, &transition_queue);

        Self {
            db: None,
            state: EntryState.into(),
            task_tx,
            task_handle,
            transition_queue,
        }
    }

    pub async fn dispatch(&mut self, event: Event) -> Result<Transition> {
        State::consume(self, event).await
    }
}
