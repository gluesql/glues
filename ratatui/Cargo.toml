[package]
name = "glues"
authors.workspace = true
version.workspace = true
edition.workspace = true
license.workspace = true
repository.workspace = true
description = "Lightweight TUI note-taking app with Git, CSV, and JSON support - privacy-focused and sync-enabled"
readme = "../README.md"
keywords = ["tui", "note-taking", "notes-app", "ratatui", "data-privacy"]
default-run = "glues"

[dependencies]
glues-core.workspace = true
gluesql.workspace = true
futures = "0.3.30"
async-io = "2.3.2"
ratatui = "0.28.1"
color-eyre = "0.6.3"
tui-big-text = "0.6.0"
tui-textarea = "0.6.1"
home = "0.5.9"
