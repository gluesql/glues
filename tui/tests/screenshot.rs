use glues::{App, config, logger};
use insta::assert_snapshot;
use ratatui::{
    Terminal,
    backend::TestBackend,
    crossterm::event::{Event as Input, KeyCode, KeyEvent, KeyModifiers},
};
use std::{fs, path::Path};

fn save_buffer(backend: &TestBackend, name: &str) {
    let dir = Path::new("tui/tests/frames");
    fs::create_dir_all(dir).unwrap();
    let path = dir.join(name);
    fs::write(path, format!("{}", backend)).unwrap();
}

async fn send_key(app: &mut App, terminal: &mut Terminal<TestBackend>, code: KeyCode) {
    let input = Input::Key(KeyEvent::new(code, KeyModifiers::NONE));
    app.handle_input(input).await;
    terminal.draw(|f| app.draw(f)).unwrap();
}

#[tokio::test]
async fn scenario_screenshots() {
    config::init().await;
    logger::init().await;
    let mut app = App::new().await;
    let backend = TestBackend::new(80, 24);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|f| app.draw(f)).unwrap();
    save_buffer(terminal.backend(), "frame_0_entry.txt");

    send_key(&mut app, &mut terminal, KeyCode::Char('a')).await;
    save_buffer(terminal.backend(), "frame_1_help.txt");

    send_key(&mut app, &mut terminal, KeyCode::Char('x')).await; // close help
    send_key(&mut app, &mut terminal, KeyCode::Char('1')).await; // open memory
    save_buffer(terminal.backend(), "frame_2_notebook.txt");

    send_key(&mut app, &mut terminal, KeyCode::Char('m')).await; // directory actions
    save_buffer(terminal.backend(), "frame_3_dir_actions.txt");

    send_key(&mut app, &mut terminal, KeyCode::Enter).await; // prompt add note
    save_buffer(terminal.backend(), "frame_4_prompt.txt");

    for ch in "Hello".chars() {
        send_key(&mut app, &mut terminal, KeyCode::Char(ch)).await;
    }
    save_buffer(terminal.backend(), "frame_5_typing.txt");

    send_key(&mut app, &mut terminal, KeyCode::Enter).await; // create note
    save_buffer(terminal.backend(), "frame_6_note_added.txt");

    assert_snapshot!("scenario_final", format!("{}", terminal.backend()));
}
