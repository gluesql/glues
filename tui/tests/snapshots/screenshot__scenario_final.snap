---
source: tui/tests/screenshot.rs
assertion_line: 55
expression: "format!(\"{}\", terminal.backend())"
snapshot_kind: text
---
" Note 'Hello' selected                                         [?] Show keymap "
"[Browser]                                   ▐[Editor]                           "
" 󰝰 Notes                                    ▐ 1 Welcome to Glues :D             "
"   󱇗 Sample Note                            ▐                                   "
"   󱇗 Hello                                  ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
"                                            ▐                                   "
